File: sliding_window/permutation_in_string/js/permutation_in_string.js
const s1 = "ab";
const s2 = "eidbaooo";

const checkInclusion = (s1, s2) => {
  let index = {};
  let s = s1.split("");

  s.forEach(c => {
    let n = index[c] || 0;
    index[c] = ++n;
  });

  let [l, r] = [0, 0];

  while(r < s2.length) {
    let n = index[s2[r]];
    index[s2[r]] = n === undefined ? -1 : n - 1;

    while(index[s2[r]] < 0) {
      index[s2[l]]++;
      l++;
    }

    if(r - l + 1 === s1.length) return true;
    
    r++;
  }
  return false;
}

console.log(checkInclusion(s1, s2));

File: sliding_window/permutation_in_string/cpp/permutation_in_string.cpp
#include<iostream>
#include<cstring>

using std::string;

// Time O(n)
// Space O(1)

class Solution {
public:
    static bool checkInclusion(const string s1, const string s2) {
        int index[26]{};
        int l = 0, r = 0;
        for(const auto& c : s1) {
            index[c-'a']++;
        }

        /* Average sliding window, uses ascii indexing
         * if conditions broken, increase l until condition 
         * is true again. match size to determine correctness
         * this size match is faster than the common loop check
         */
        while(r < s2.size()) {
            index[s2[r]-'a']--;
            while(index[s2[r]-'a'] < 0) {
                index[s2[l]-'a']++;
                l++;
            }
            if(r - l + 1 == s1.size()) {
                return true;
            }
            r++;
        }
        return false;
    }
};

int main() {
    string s1 = "abcdxabcde";
    string s2 = "abcdeabcdx";

    bool answer = Solution::checkInclusion(s1, s2);
    std::cout << std::boolalpha <<  answer << '\n';

    return 0;
}
